### ISSUE TITLE
About the LRW1000 experiment

### ISSUE THREAD

**Issue by Grit1024** [ms-dot-k/Multi-head-Visual-Audio-Memory#5]
Hi，thanks for your effective work. 
I'm trying to reproduce this experiment on the LRW1000, but the specific settings of some parameters are bothering me, and I was wondering if you could provide specific settings to help get the best performance on this dataset?
Thanks！

**Comment by dydtlsqj** [comment]
Hello, for LRW 1000, you need to write your own dataset script.
Especially, LRW 1000 has different time lengths for each video, you should consider it by writing your own collate_fn (zero pad to have the same video length in a batch).
Also, please note that the audio provided by LRW1000 is 0.4 longer than provided word boundary for the video. So we experimented by using longer video by adding extra 0.2 sec frames in both back and forth.
The other parameters like learning rate, memory size, etc are the same with the LRW experiment except for the batch size.

**Comment by Grit1024** [comment]
> 

Yes, I noticed the collate function in vid_aud_lrw1000.py, but I don't see any trace of the program using it, can you roughly explain it?

**Comment by Grit1024** [comment]
> Hello, for LRW 1000, you need to write your own dataset script. Especially, LRW 1000 has different time lengths for each video, you should consider it by writing your own collate_fn (zero pad to have the same video length in a batch). Also, please note that the audio provided by LRW1000 is 0.4 longer than provided word boundary for the video. So we experimented by using longer video by adding extra 0.2 sec frames in both back and forth. The other parameters like learning rate, memory size, etc are the same with the LRW experiment except for the batch size.

Yes, I noticed the collate function in vid_aud_lrw1000.py, but I don't see any trace of the program using it, can you roughly explain it?

**Comment by ms-dot-k** [comment]
The collate function can be utilized at dataloader in train() and validate() as follows:
```
        dataloader = DataLoader(
        train_data,
        shuffle=False if args.distributed else True,
        batch_size=args.batch_size,
        num_workers=args.workers,
        collate_fn=lambda x: train_data.collate(x),
        pin_memory=True,
        sampler=train_sampler
    )
```

**Comment by Grit1024** [comment]
> The collate function can be utilized at dataloader in train() and validate() as follows:
> 
> ```
>         dataloader = DataLoader(
>         train_data,
>         shuffle=False if args.distributed else True,
>         batch_size=args.batch_size,
>         num_workers=args.workers,
>         collate_fn=lambda x: train_data.collate(x),
>         pin_memory=True,
>         sampler=train_sampler
>     )
> ```

Thank you very much for your reply, it's very helpful, but I'm having problems with the` transform(aud)` stage where `mat1 and mat2 shapes cannot be multiplied  (201x1 and 201x80)`, I think I need to go into more detail.

In addition, I would like to know for loading audio and video data, does it have to be that audio and video correspond to the same label, e.g. (video, audio, word), and is it possible to load audio and video data separately and simultaneously for training of different networks?

Sorry I'm still learning about this, so if you can point me in the right direction, thank you very much!

**Comment by ms-dot-k** [comment]
The problem may be due to the version mismatch of torchaudio.
Try to change lines #148~
```
torchaudio.transforms.Spectrogram(win_length=400, hop_length=160, power=None, normalized=True), # 100 fps (hop_length 10ms)
torchaudio.transforms.ComplexNorm(2),
torchaudio.transforms.MelScale(n_mels=80, sample_rate=16000),
```
into one line
```
torchaudio.transforms.MelSpectogram(win_length=400, hop_length=160, normalized=True ,n_mels=80, sample_rate=16000)
```

The data loader loads audio, video, and text (word class index). So you can pass the audio and video into different networks.

**Comment by Grit1024** [comment]
I modified MelSpectrogram according to your suggestion and it solved the problem, which is great.

In the collate function I tried to change `vid_length = [data[3] for data in batch if data[5]! = 1]` to `vid_length = [data[3] for data in batch if data[5] == 1]`, because the former gives the error that vid_length is an empty list, and should `data[5]` be referred to as the `skip` parameter? I don't quite understand what this parameter represents.

Also, I realized that I end up `return spec, vid, vid_length, targets`, but `vid_length` doesn't seem to behave in the train, so I just modified `a_in, v_in, _, target = batch`, and I'd like to know if there are any more refinements to the `train.py` module. I'd like to know if there are more improvements to be made to the train.py module.

Thank you very much!

**Comment by ms-dot-k** [comment]
When the skip parameter is set to 1, it indicates the loaded data is unnormal. Therefore, we shouldn't use them when making a batch.

Finally, the vid_length should be utilized when training lrw-1000 dataset.
We make the final prediction by average pooling the whole sequence for LRW.
However, for the lrw-1000, you should only take account into the valid video length only. (By not averaging the padded sequences)

**Comment by Grit1024** [comment]
Thank you for your response, I think I have a basic understanding of this experimental idea and it is a valuable piece of work. 
Best wishes!
